go/vt/vtgr/controller/diagnose.go

```diff
@@ -0,0 +1,561 @@
+		return ""
+	}
+	return shard.sqlGroup.GetGroupName()
+}
+
+func (shard *GRShard) isAllOfflineOrError() bool {
+	return shard.sqlGroup.IsAllOfflineOrError()
+}
+
+func (shard *GRShard) getOnlineGroupInfo() (int, bool) {
+	return shard.sqlGroup.GetOnlineGroupInfo()
+}
+
+func (shard *GRShard) hasWrongPrimaryTablet(ctx context.Context) (bool, error) {
+	// Find out the hostname and port of the primary in mysql group
+	// we try to use local instance and then fallback to a random instance to check mysqld
+	// in case the primary is unreachable
+	host, port, _ := shard.sqlGroup.GetPrimary()
+	if !isHostPortValid(host, port) {
+		log.Warningf("Invalid address for primary %v:%v", host, port)
+		return false, errMissingGroup
+	}
+	// Make sure we have a tablet available
+	// findTabletByHostAndPort returns nil when we cannot find a tablet
+	// that is running on host:port, which means the tablet get stuck
+	// or when the tablet is not reachable
+	// we retrun errMissingPrimaryTablet so that VTGR will trigger a failover
+	tablet := shard.findTabletByHostAndPort(host, port)
+	if tablet == nil || !shard.instanceReachable(ctx, tablet) {
+		log.Errorf("Failed to find tablet that is running with mysql on %v:%v", host, port)
+		return false, errMissingPrimaryTablet
+	}
+	// Now we know we have a valid mysql primary in the group
+	// we should make sure tablets are aligned with it
+	primary := shard.findShardPrimaryTablet()
+	// If we failed to find primary for shard, it mostly means we are initializing the shard
+	// return true directly so that VTGR will set primary tablet according to MySQL group
+	if primary == nil {
+		log.Infof("unable to find primary tablet for %v", formatKeyspaceShard(shard.KeyspaceShard))
+		return true, nil
+	}
+	return (host != primary.instanceKey.Hostname) || (port != primary.instanceKey.Port), nil
+}
+
+func (shard *GRShard) isPrimaryReachable(ctx context.Context) (bool, error) {
+	primaryTablet := shard.findShardPrimaryTablet()
+	if primaryTablet == nil {
+		return false, fmt.Errorf("unable to find primary for %v", formatKeyspaceShard(shard.KeyspaceShard))
+	}
+	return shard.instanceReachable(ctx, primaryTablet), nil
+}
+
+func (shard *GRShard) instanceReachable(ctx context.Context, instance *grInstance) bool {
+	pingCtx, cancel := context.WithTimeout(context.Background(), *pingTabletTimeout)
+	defer cancel()
+	c := make(chan error, 1)
+	// tmc.Ping create grpc client connection first without timeout via dial
+	// then call the grpc endpoint using the context with timeout
+	// this is problematic if the host is really unreachable, we have to wait the
+	// all the retries inside grpc.dial with exponential backoff
+	go func() { c <- shard.tmc.Ping(pingCtx, instance.tablet) }()
+	select {
+	case <-pingCtx.Done():
+		log.Errorf("Ping abort timeout %v", *pingTabletTimeout)
+		return false
+	case err := <-c:
+		if err != nil {
+			log.Errorf("Ping error host=%v: %v", instance.instanceKey.Hostname, err)
+		}
+		return err == nil
+	}
```

Comment:
This is bad but it's actually our fault. Once https://github.com/vitessio/vitess/pull/8368 is merged we should be able tof ix the TMC API so that the initial dial is context-aware and we don't need this hack.

Reply:
I don't think switching to DialContext there will actually fix this issue, though. From [the docs](https://pkg.go.dev/google.golang.org/grpc#DialContext)

> By default, it's a non-blocking dial (the function won't wait for connections to be established, and connecting happens in the background). To make it a blocking dial, use WithBlock() dial option.
>
> In the non-blocking case, the ctx does not act against the connection. It only controls the setup steps.

I don't think we should unilaterally use `WithBlock` as that will result in slowdowns in the rest of the tmc use-cases, so what I think we should do is a follow-up refactor to the tmc api (and merge both #8368 (pending review / further changes) and this PR as-is) to allow callers to specify dial options to use, something like:

```go
tmc := tmclient.TabletManagerClient().WithDialOptions(grpc.WithBlock())
```

Reply:
I think this sounds about right: the comment that @narcsfz added to this method was a bit confusing, because it complains about dial/retry timeouts when (in theory) we shouldn't be seeing dial timeouts at all, since this API is not called with `WithBlock` and hence the actual connection will only happen during the actual call to `Ping`, and will be context-aware. I think what @narcsfz was seeing are the basic setup steps which by default are not controlled by context but _will be_ if we switch to `DialContext`.

Reply:
Thanks @vmg @ajm188. I'm happy to leave a TODO here and refactor it later if that makes sense. Since it is hard to reproduce in unit test, I also need to set up the test case that can reproduce the problem to test the change out.